stl/inc/format

```diff
@@ -366,58 +366,347 @@ private:
+        return static_cast<long long>(42);
+    } else if constexpr (unsigned_integral<_Ty> && sizeof(_Ty) <= sizeof(unsigned int long long)) {
+        return static_cast<unsigned int long long>(42);
+    } else if constexpr (is_same_v<_Ty, bool>) {
+        return bool{};
+    } else if constexpr (is_same_v<_Ty, float>) {
+        return float{};
+    } else if constexpr (is_same_v<_Ty, double>) {
+        return double{};
+    } else if constexpr (is_same_v<_Ty, long double>) {
+        return static_cast<long double>(42);
+    } else if constexpr (is_same_v<_Ty, const void*>) {
+        return static_cast<const void*>(nullptr);
+    } else if constexpr (is_same_v<_Ty, const _CharType*>) {
+        return static_cast<const _CharType*>(nullptr);
+    } else if constexpr (is_same_v<_Ty, basic_string_view<_CharType>>) {
+        return basic_string_view<_CharType>{};
+    } else {
+        return basic_format_arg<_Context>::handle();
+    }
+}
+
+template <class _Context, class _Ty>
+/* consteval */ constexpr size_t _Get_format_arg_storage_size() noexcept {
+    return sizeof(_Get_format_arg_storage_type<_Context, _Ty>());
+}
+
+struct _Format_arg_store_packed_index {
+    // TRANSITION, Should be templated on number of arguments for even less storage
+    using _Index_type = size_t;
+
+    constexpr _Format_arg_store_packed_index() = default;
+    explicit constexpr _Format_arg_store_packed_index(const size_t _Index)
+        : _Index(static_cast<_Index_type>(_Index)), _Type(_Basic_format_arg_type::_None) {}
+
+    _Index_type _Index : (sizeof(_Index_type) * 8 - 4);
+    _Basic_format_arg_type _Type : 4;
 };
 
-// TODO: test coverage
 template <class _Context, class... _Args>
 class _Format_arg_store {
-    static constexpr size_t _Num_args = sizeof...(_Args);
-    // TODO: no arg packing yet
+private:
+    template <class>
+    friend class basic_format_args;
+
+    static constexpr size_t _Num_args       = sizeof...(_Args);
+    static constexpr size_t _Storage_length = (_Get_format_arg_storage_size<_Context, _Args>() + ... + 0);
+
+    using _CharType   = typename _Context::char_type;
+    using _Index_type = _Format_arg_store_packed_index;
+
+    template <class _Ty, _Basic_format_arg_type _Arg_type>
+    void _Store_impl(const size_t _Arg_index, _Ty _Val) noexcept {
+        const auto _Store_index = _Index_array[_Arg_index]._Index;
+        const auto _Length      = _Get_format_arg_storage_size<_Context, _Ty>();
+
+        _CSTD memcpy(_Storage + _Store_index, _STD addressof(_Val), _Length);
+        _Index_array[_Arg_index]._Type = _Arg_type;
+        if (_Arg_index + 1 < _Num_args) { // Set the starting index of the next arg, as that is dynamic
+            _Index_array[_Arg_index + 1] = _Format_arg_store_packed_index{_Store_index + _Length};
+        }
+    }
+
+    byte _Storage[(_STD max)(_Storage_length, static_cast<size_t>(1))];
+    _Index_type _Index_array[(_STD max)(_Num_args, static_cast<size_t>(1))];

```

Comment:
Would it make sense to specialize `_Format_arg_store` for empty `_Args` so we can avoid silliness like these `max`s?

Reply:
doesn't std::array have such a specialization already? we could just use that.

Reply:
I mean we are already pulling in half the world anyway. 

I will try a specialization with 0 arguments after I have reworked the storage

Reply:
It does, but we weren't using `array`, and specializing this class is stupid simple since it doesn't need data members or `_Store` functions. :shrug:

Reply:
I think we would need something to pass to basic_format_args, although we could specialize that constructor too

Reply:
why? basic_format_args should be default constructable

Reply:
I meant the constructor that takes an empty `_Format_args_storage`

Reply:
doesn't the standard specify that "getting" the first format arg from  format_args initialized from empty storage should return a "none/monostate" format_arg anyways.

Given that I'm pretty OK with just implementing that by ensuring the arrays are at least 1 sized.

Reply:
Implemented the specialization after a bit of refactoring